shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/NativeRequestContext.cs

```diff
@@ -0,0 +1,458 @@
+            if (pKnownHeader->pRawValue != null)
+            {
+                value = HeaderEncoding.GetString(pKnownHeader->pRawValue + fixup, pKnownHeader->RawValueLength);
+            }
+
+            return value;
+        }
+
+        internal void GetUnknownHeaders(IDictionary<string, StringValues> unknownHeaders)
+        {
+            if (_perminatelyPinned)
+            {
+                GetUnknownHeadersHelper(unknownHeaders, 0, _nativeRequest);
+            }
+            else
+            {
+                // Return value.
+                fixed (byte* pMemoryBlob = _backingBuffer)
+                {
+                    var request = (HttpApiTypes.HTTP_REQUEST*)(pMemoryBlob + _bufferAlignment);
+                    long fixup = pMemoryBlob - (byte*)_originalBufferAddress;
+                    GetUnknownHeadersHelper(unknownHeaders, fixup, request);
+                }
+            }
+        }
+
+        private void GetUnknownHeadersHelper(IDictionary<string, StringValues> unknownHeaders, long fixup, HttpApiTypes.HTTP_REQUEST* request)
+        {
+            int index;
+
+            // unknown headers
+            if (request->Headers.UnknownHeaderCount != 0)
+            {
+                var pUnknownHeader = (HttpApiTypes.HTTP_UNKNOWN_HEADER*)(fixup + (byte*)request->Headers.pUnknownHeaders);
+                for (index = 0; index < request->Headers.UnknownHeaderCount; index++)
+                {
+                    // For unknown headers, when header value is empty, RawValueLength will be 0 and
+                    // pRawValue will be null.
+                    if (pUnknownHeader->pName != null && pUnknownHeader->NameLength > 0)
+                    {
+                        var headerName = HeaderEncoding.GetString(pUnknownHeader->pName + fixup, pUnknownHeader->NameLength);
+                        string headerValue;
+                        if (pUnknownHeader->pRawValue != null && pUnknownHeader->RawValueLength > 0)
+                        {
+                            headerValue = HeaderEncoding.GetString(pUnknownHeader->pRawValue + fixup, pUnknownHeader->RawValueLength);
+                        }
+                        else
+                        {
+                            headerValue = string.Empty;
+                        }
+                        // Note that Http.Sys currently collapses all headers of the same name to a single coma separated string,
+                        // so we can just call Set.
+                        unknownHeaders[headerName] = headerValue;
+                    }
+                    pUnknownHeader++;
+                }
+            }
+        }
+
+        internal SocketAddress GetRemoteEndPoint()
+        {
+            return GetEndPoint(localEndpoint: false);
+        }
+
+        internal SocketAddress GetLocalEndPoint()
+        {
+            return GetEndPoint(localEndpoint: true);
+        }
+
+        private SocketAddress GetEndPoint(bool localEndpoint)
```

Comment:
No helper method for this one to reduce redundancy?

Reply:
Because of https://github.com/aspnet/HttpSysServer/pull/397/files/7c75482b9c69488c871f66a7322faec321bf1b69#diff-836eec03ef1919362c7be39f37747f57R356, I can't refactor the rest of the method completely out, but I can make partial helper.

Reply:
You probably can factor that line out, you'd just have to pass in pMemoryBlob, which is the same as `(IntPtr)HttpApiTypes.HTTP_REQUEST*` for IIS.

Reply:
I think I can pass in 0 for pMemoryBlob for the IIS case.